mitiq/shadows/quantum_processing.py

```diff
@@ -0,0 +1,181 @@
+        assert len(pauli_string) == num_qubits, (
+            f"Pauli string must be same length as number of qubits, "
+            f"got {len(pauli_string)} and {num_qubits}"
+        )
+        rotated_circuit = circuit.copy()
+        for i, pauli in enumerate(pauli_string):
+            qubit = qubits[i]
+            # Pauli X measurement is equivalent to H plus a Z measurement
+            if pauli == "X":
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli X measurement is equivalent to S^-1*H plus a Z measurement
+            elif pauli == "Y":
+                rotated_circuit.append(cirq.S(qubit) ** -1)
+                rotated_circuit.append(cirq.H(qubit))
+            # Pauli Z measurement
+            else:
+                assert (
+                    pauli == "Z"
+                ), f"Pauli must be X, Y, Z. Got {pauli} instead."
+        if add_measurements:
+            # append measurement gates
+            rotated_circuit.append(cirq.measure(*qubits))
+        # append to list of circuits
+        rotated_circuits.append(rotated_circuit)
+    return rotated_circuits
+
+
+# Stage 1 of Classical Shadows: Measurements
+def random_pauli_basis_measurement(
+    circuit: cirq.Circuit,
+    n_total_measurements: int,
+    sampling_function: Union[str, Callable[..., MeasurementResult]] = "cirq",
+    sampling_function_config: Dict[str, Any] = {},
+) -> Tuple[NDArray[Any], NDArray[Any]]:
+    r"""Given a circuit, perform random Pauli-basis measurements on
+    the circuit and return the measurement outcomes & the sampled
+    Pauli strings.
+
+    Args:
+         circuit: Cirq circuit.
+         n_total_measurements: number of snapshots.
+         sampling_function: Sampling function to use. If a string, then the
+            string is used to look up a default sampling function. If a
+            Callable, then the Callable is used as the sampling function.
+            Defaults to `"cirq"`.
+         sampling_function_config: Configuration
+            for the sampling function. Defaults to {}. If sampling_function is
+            a string, then this argument is ignored.
+
+    Returns:
+         outcomes: Tuple of two numpy arrays. The first array contains
+            measurement outcomes (-1, 1) while the second array contains the
+            index for the sampled Pauli's (0,1,2=X,Y,Z). Each row of the arrays
+            corresponds to a distinct snapshot or sample while each column
+            corresponds to a different qubit.
+    """
+
+    # Generate random Pauli unitaries
+    qubits = list(circuit.all_qubits())
+    num_qubits = len(qubits)
+    pauli_strings = generate_random_pauli_strings(
+        num_qubits, n_total_measurements
+    )
+    # Attach measurement gates to the circuit
+    rotated_circuits = tqdm(
+        get_rotated_circuits(
+            circuit,
+            pauli_strings,
+        ),
+        desc="Measurement",
+        leave=False,
+    )
```

Comment:
Does this take a long time? Wondering if `tqdm` is needed here.

Reply:
yes, it takes a long time when measurements get larger.  I think it's important to keep track of the time before killing the program. But I have modified it to an optional one.